packages/eslint-plugin/src/rules/analytics-configuration.ts

```diff
@@ -0,0 +1,217 @@
+import { ESLintUtils, TSESTree } from "@typescript-eslint/utils"
+
+type Options = readonly [
+  {
+    reservedPropNames: string[]
+  }
+]
+
+const messages = {
+  reservedPayloadPropNames: "{{ propName }} is a reserved payload prop name.",
+
+  eventNameFormat:
+    "Event name {{ eventName }} must be in SCREAMING_SNAKE_CASE.",
+  emptyPayloadType:
+    "Events with empty payloads must use `never` for the payload type.",
+  invalidPayloadFormat:
+    "Payloads must be specific key-value pairs where keys are strings and values are either string, boolean, or number.",
+} as const
+
+type MessageIds = keyof typeof messages
+
+/**
+ * Require event names to:
+ * - Be all caps
+ * - Start and end with a letter
+ * - Only use underscore to delimit words
+ */
+const eventNameRegex = /^[A-Z][A-Z_]*[A-Z]$/
+
+const isTypeAnnotationEmpty = (node: TSESTree.TSPropertySignature): boolean => {
+  const typeAnnotation = node.typeAnnotation as TSESTree.TSTypeAnnotation
+  const innerTypeAnnotation = typeAnnotation.typeAnnotation
+
+  if (
+    innerTypeAnnotation.type === "TSTypeLiteral" &&
+    innerTypeAnnotation.members.length === 0
+  ) {
+    // Disallow empty objects
+    return true
+  }
+
+  if (
+    innerTypeAnnotation.type === "TSNullKeyword" ||
+    innerTypeAnnotation.type === "TSUndefinedKeyword"
+  ) {
+    return true
+  }
+
+  return false
+}
+
+const getPropertyName = (node: TSESTree.TSPropertySignature): string => {
+  if (node.key.type === "Identifier") {
+    return node.key.name
+  }
+
+  if (node.key.type === "Literal" && node.key.value) {
+    return node.key.value?.toString()
+  }
+
+  throw new Error(
+    `Could not identify event name on node type: ${node.key.type}`
+  )
+}
+
+// Use without docs for now... should we add a section in the reference
+// documentation explaining custom rules?
```

Comment:
Assuming this refers to external docs urls, like https://eslint.org/docs/latest/rules/no-const-assign, for example? I think it's okay to go without for now, but yes this would be a nice addition to the docs site, even if it was a single page with jump links for each rule.

Reply:
Yup, that's exactly it.

Reply:
Actually, it's easy enough to add a doc site page for these, I'll go ahead and do so.